### PR TITLE
8267663: [vector] Add unsigned comparison operators on AArch64

### PR DESCRIPTION
**Pull request by theRealELiu**
This patch implements unsigned vector comparison on AArch64. The
performance of unsigned comparison improves about 4x~5x in my local with
Byte128Vector.java[1].

Before:
Benchmark                               Score(op/ms)     Error
Byte128Vector.UNSIGNED_GE#size(1024)    99.953           6.17
Byte128Vector.UNSIGNED_GT#size(1024)    95.334           8.865
Byte128Vector.UNSIGNED_LE#size(1024)    76.908           24.332
Byte128Vector.UNSIGNED_LT#size(1024)    78.362           23.507

After:
Benchmark                               Score(op/ms)     Error
Byte128Vector.UNSIGNED_GE#size(1024)    421.809          25.57
Byte128Vector.UNSIGNED_GT#size(1024)    420.653          26.779
Byte128Vector.UNSIGNED_LE#size(1024)    316.754          92.889
Byte128Vector.UNSIGNED_LT#size(1024)    423.683          26.508

[Test]
- All vector API test cases passed without new failure. 8265312[2] has
  been implemented this on x86 and supplied sufficient test cases for
  all kinds of vector.
- No performance regression for other comparisons.
- libjvm.so drops off about 200KB after this patch by combining those
  vector compare rules.

[1] https://github.com/openjdk/panama-vector/blob/vectorIntrinsics/test/jdk/jdk/incubator/vector/benchmark/src/main/java/benchmark/jdk/incubator/vector/Byte128Vector.java#L1198
[2] https://github.com/openjdk/panama-vector/pull/68

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267663](https://bugs.openjdk.java.net/browse/JDK-8267663): [vector] Add unsigned comparison operators on AArch64

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4358/head:pull/4358` \
`$ git checkout pull/4358`

Update a local copy of the PR: \
`$ git checkout pull/4358` \
`$ git pull https://git.openjdk.java.net/jdk pull/4358/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4358`

View PR using the GUI difftool: \
`$ git pr show -t 4358`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4358.diff">https://git.openjdk.java.net/jdk/pull/4358.diff</a>

</details>

### CODE REVIEW

**Comment by adinn** [comment]
The title of this PR and the JBS issue need changing to reflect the fact that this is adding unsigned comparison operators for *vectors* on AArch64.

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4 [hunk 1]
```diff
@@ -864,119 +864,97 @@ VECTOR_EXTRACT_F(2, D, X, D)
...
+      return Assembler::T2D;
+    default:
+      assert(false, "unsupported");
+      ShouldNotReachHere();
+  }
+  return Assembler::INVALID_ARRANGEMENT;
+}
+
+static void neon_compare(C2_MacroAssembler masm, FloatRegister dst, BasicType bt,
+                         FloatRegister src1, FloatRegister src2, int cond, bool isX) {
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:889]: Passing masm by value, rather than reference, is unconventional. There's no reason to do that.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:889]: Come to think about it, why isn't this an Assembler method?

Comment (theRealELiu) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4]: The latest fix moves it into macroAssembler, it's more reasonable.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 2]
```diff
@@ -5339,6 +5339,66 @@ void MacroAssembler::safepoint_isb() {
 #endif
 }
 
+Assembler::SIMD_Arrangement elemBytes_to_Arrangement(int esize, bool isQ) {
+  switch(esize) {
+    case 1:
+      return isQ ? Assembler::T16B : Assembler::T8B;
+    case 2:
+      return isQ ? Assembler::T8H : Assembler::T4H;
+    case 4:
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5348]: This should be

`
      return isQ ? T8H : T4H;
`
but isn't this just a table lookup?

```
if (in_range)
    return elemBytes_to_Arrangement_table[esize][isQ]
```
Why not?

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 3]
```diff
@@ -5339,6 +5339,66 @@ void MacroAssembler::safepoint_isb() {
...
+      case BoolTest::ne: {
+        fcmeq(dst, size, src1, src2);
+        notr(dst, T16B, dst);
+        break;
+      }
+      case BoolTest::ge: fcmge(dst, size, src1, src2); break;
+      case BoolTest::gt: fcmgt(dst, size, src1, src2); break;
+      case BoolTest::le: fcmge(dst, size, src2, src1); break;
+      case BoolTest::lt: fcmgt(dst, size, src2, src1); break;
+      default:
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5374]: Are you sure about these two lines?

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 4]
```diff
@@ -5339,6 +5339,66 @@ void MacroAssembler::safepoint_isb() {
...
+      case BoolTest::eq: fcmeq(dst, size, src1, src2); break;
+      case BoolTest::ne: {
+        fcmeq(dst, size, src1, src2);
+        notr(dst, T16B, dst);
+        break;
+      }
+      case BoolTest::ge: fcmge(dst, size, src1, src2); break;
+      case BoolTest::gt: fcmgt(dst, size, src1, src2); break;
+      case BoolTest::le: fcmge(dst, size, src2, src1); break;
+      case BoolTest::lt: fcmgt(dst, size, src2, src1); break;
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5373]: Are you sure about these two lines?

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 5]
```diff
@@ -5339,6 +5339,66 @@ void MacroAssembler::safepoint_isb() {
...
+      case BoolTest::eq: cmeq(dst, size, src1, src2); break;
+      case BoolTest::ne: {
+        cmeq(dst, size, src1, src2);
+        notr(dst, T16B, dst);
+        break;
+      }
+      case BoolTest::ge: cmge(dst, size, src1, src2); break;
+      case BoolTest::gt: cmgt(dst, size, src1, src2); break;
+      case BoolTest::le: cmge(dst, size, src2, src1); break;
+      case BoolTest::lt: cmgt(dst, size, src2, src1); break;
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5389]: And these two?

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp (hunk 4 above)
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5373]: Sorry, I see what's happening: the operands are reversed. This is tricky code.

Comment (theRealELiu) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5356]: As A64-SIMD doesn't have `fcmlt(register)` and `fcmle(register)` instructions, I think it's okay to use `fcmgt` and `fcmge` with reversing the argument. But I can't locate the initial patch adding them.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp (hunk 2 above)
Comment (theRealELiu) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp]: Thanks. The table is done.

File: src/hotspot/cpu/aarch64/assembler_aarch64.cpp [hunk 6]
```diff
@@ -41,6 +41,25 @@ short Assembler::SIMD_Size_in_bytes[] = {
...
+  /*   7  */      {INVALID_ARRANGEMENT, INVALID_ARRANGEMENT},
+  /*   8  */      {T1D,                 T2D}
+  };
+
+
+Assembler::SIMD_Arrangement Assembler::esize2arrangement(int esize, bool isQ) {
+    guarantee(esize == 1 || esize == 2 || esize == 4 || esize == 8, "unsupported element size");
+    return _esize2arrangement_table[esize][isQ];
+}
+
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.cpp:62]: That's nice.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp (hunk 4 above)
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5356]: Yes, I see now.

**Review by theRealAph — APPROVED** [review]
OK, thanks. That's a great improvement.

**Pull request closed without merging**